src/System.Linq/tests/WhereTests.cs

```diff
@@ -0,0 +1,644 @@
+
+            Assert.Equal(0, result.Count());
+        }
+
+        [Fact]
+        public void Where_EmptyEnumerable_ReturnsNoElements()
+        {
+            IEnumerable<int> source = Enumerable.Empty<int>();
+            bool wasSelectorCalled = false;
+
+            IEnumerable<int> result = source.Where(value => { wasSelectorCalled = true; return true; });
+            
+            Assert.Equal(0, result.Count());
+            Assert.False(wasSelectorCalled);
+        }
+        
+        [Fact]
+        public void Where_Array_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            int[] source = new[] { 1 };
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+        
+        [Fact]
+        public void Where_List_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            List<int> source = new List<int>() { 1 };
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void Where_IReadOnlyCollection_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            IReadOnlyCollection<int> source = new ReadOnlyCollection<int>(new List<int>() { 1 });
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void Where_ICollection_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            ICollection<int> source = new LinkedList<int>(new List<int>() { 1 });
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void Where_IEnumerable_CurrentIsDefaultOfTAfterEnumeration()
+        {
+            IEnumerable<int> source = Enumerable.Repeat(1, 1);
+            Func<int, bool> truePredicate = (value) => true;
+
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void WhereWhere_Array_ReturnsExpectedValues()
```

Comment:
The "WhereWhere" combination is interesting because Where creates a custom Iterator for Where which is then recognized by subsequent Where calls.  Potentially more interesting though is the "WhereSelect" combination, as there are special code paths to handle that common pattern.  Would you consider adding some tests for that combination, too?

There's also one other Where overload, that takes a Func<TSource, int, bool> instead of a Func<TSource, bool>.  Do you have any plans to add tests for that overload, too?